### PR TITLE
Fix PR creation via api between branches of same repo with head field namespaced

### LINKED ISSUE #20175: POSTing to `/repos/{owner}/{repo}/pulls` with explicit head incorrectly returns 404 if base and head are the same repository
## Observed

When POSTing to `/repos/{owner}/{repo}/pulls` to create a pull request
with a payload which specifies `head` to be the same repository as in
the request such that the source and the target repository are the
same and not a fork, the REST API returns a 404 with an error message
`GetForkedRepo`.

e.g. in route `myuser/repo` and `myuser:some-branch` in payload.

## Expected

I expect this to create the pull request instead of just returning an
incorrect and confusing error message.

# Details

The bug seems to be that the branch in
https://github.com/go-gitea/gitea/blob/main/routers/api/v1/repo/pull.go#L940-L951
does not check if it actually is the same repository and thus doesn't
set `isSameRepo` to true.

## Versions

I saw this behaviour when runnig gitea in a FreeBSD 13.1-RELEASE jail:

```console
# jexec gitea freebsd-version
13.1-RELEASE
# jexec gitea uname -a
FreeBSD gitea 13.1-RELEASE FreeBSD 13.1-RELEASE releng/13.1-n250148-fc952ac2212 TRITON131 amd64
#
#
# pkg -j gitea info gitea
gitea-1.16.7
Name           : gitea
Version        : 1.16.7
Installed on   : Sat Jun 18 15:48:24 2022 CEST
Origin         : www/gitea
Architecture   : FreeBSD:13:amd64
Prefix         : /usr/local
Categories     : www
Licenses       : MIT
Maintainer     : stb@lassitu.de
WWW            : https://gitea.io/en-US/
Comment        : Compact self-hosted Git service
Options        :
	BINDATA        : off
	GIT_LFS        : on
	PAM            : on
	SQLITE         : on
Annotations    :
	FreeBSD_version: 1300139
	cpe            : cpe:2.3:a:gitea:gitea:1.16.7:::::freebsd13:x64
	repo_type      : binary
	repository     : FreeBSD
Flat size      : 104MiB
Description    :
A self-hosted Git service.

Gitea is a community managed fork of Gogs, lightweight code hosting solution
written in Go and published under the MIT license.

WWW: https://gitea.io/en-US/
#
```

### PR DESCRIPTION
**Pull request by norohind**
Fix #20175

Current implementation of API does not allow creating pull requests between branches of the same
repo when you specify *namespace* (owner of the repo) in `head` field in http request body.

--- 

Although GitHub implementation of API allows performing such action and since Gitea targeting
compatibility with GitHub API I see it as an appropriate change.

I'm proposing a fix to the described problem and test case which covers this logic.

My use-case just in case: https://github.com/go-gitea/gitea/issues/20175#issuecomment-1711283022

### CODE REVIEW

File: routers/api/v1/repo/pull.go [hunk 1]
```diff
@@ -961,6 +961,8 @@ func parseCompareInfo(ctx *context.APIContext, form api.CreatePullRequestOption)
 			return nil, nil, nil, nil, "", ""
 		}
 		headBranch = headInfos[1]
+		// The head repository can also point to the same repo
+		isSameRepo = ctx.Repo.Owner.ID == headUser.ID
```
Comment (lunny) [inline routers/api/v1/repo/pull.go:1072]: The pre-condition is one user cannot create a fork under his name.

Comment (norohind) [reply routers/api/v1/repo/pull.go:1072]: I'm sorry, I think I do not understand what does it mean in the context of this change.

**Comment by choelzl** [comment]
Sorry for the ping/bump, having automated PR creation in various projects (release, deployment, dependency, syntax,...) it would be nice seeing it resolved.

**Review by lunny — APPROVED** [review]

**Comment by lunny** [comment]
last call for this one. @go-gitea/technical-oversight-committee

**Review by wolfogre — APPROVED** [review]

**Comment by yardenshoham** [comment]
Tests are failing

**Pull request merged**